### PR TITLE
Add WebIDL sequence return values

### PR DESCRIPTION
**Pull request by emilio**
And use it for `WebGLRenderingContext::getSupportedExtensions`.
Part of #544

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8412)

### CODE REVIEW

File: components/script/dom/bindings/conversions.rs [hunk 1]
```diff
@@ -806,6 +807,28 @@ impl<T: FromJSValConvertible> FromJSValConvertible for Option<T> {
     }
 }
 
+impl<T: ToJSValConvertible> ToJSValConvertible for Vec<T> {
+    fn to_jsval(&self, cx: *mut JSContext, rval: MutableHandleValue) {
+        unsafe {
+            let js_objects: Vec<JSVal> = self.iter().map(|obj| {
+                let mut val = mem::uninitialized();
```
Comment (ecoal95) [inline components/script/dom/bindings/conversions.rs]: Not 100% sure about if this is correct, I tested it and it worked though.

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 


---

Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
There are a couple unsafe things about, this, unfortunately. If `obj.to_jsval` triggers a GC, all of the previous elements that have been been will become dangling pointers since they're not rooted. We should do this instead:

```rust
let mut js_objects = RootedVec::new();
for obj in &self {
    let mut val = RootedValue::new(cx, UndefinedValue());
    obj.to_jsval(cx, val.handle_mut());
    js_objects.push(val.handle().get());
}
```
This will ensure that each converted value is stored in a rooted vector, so subsequent conversions will not risk GCing them.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 827 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bpCFpgEQ_kKCVorjm)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L827)):</sup>
ObjectValue, since we've already checked the presence of null.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 814 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bgvafxycDLLc0CgSs-r3-814)** ([raw file](https://github.com/servo/servo/blob/82c243a63f6d99e43885636b22c4f99ca89b24d7/components/script/dom/bindings/conversions.rs#L814)):</sup>
See below :) from_marked_location is a huge red flag, in general, since it means "I swear that the address I'm passing here is known to the GC".


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by ecoal95** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
Trying this I get an error (`JSVal` is not traceable). I guess I can use a plain `Vec<RootedValue>` (the values are rooted so there should be no problem), and later get the handles?

```rust
            let js_objects: Vec<RootedValue> = self.iter().map(|obj| {                      
                let mut val = RootedValue::new(cx, UndefinedValue());
                obj.to_jsval(cx, val.handle_mut());
                val
            }).collect();                                                                   
            
            let values: Vec<JSVal> = js_objects.iter()                                      
                                               .map(|obj| obj.handle().get())               
                                               .collect(); 
```

I think this is correct (even though I'm not too happy with the double vector, but I guess we should keep the `RootedValues` while using the `JSValue`s), am I right?


---

<sup>**[components/script/dom/bindings/conversions.rs, line 814 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bgvafxycDLLc0CgSs-r3-814)** ([raw file](https://github.com/servo/servo/blob/82c243a63f6d99e43885636b22c4f99ca89b24d7/components/script/dom/bindings/conversions.rs#L814)):</sup>
Thanks!
I supposed there would be issues with GC, I'm not familiar with the SpiderMonkey API.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by ecoal95** [comment]
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
My fault, this crashes.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 827 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bpCFpgEQ_kKCVorjm)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L827)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by ecoal95** [comment]
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
I ended up using `Heap<JSValue>`, and it works fine (passes the test adding a few strings to `getSupportedExtensions`).

May yo review it to see if there might be any issue with GC, or if this can be done in a more performant way? :)

TL;DR; here is the relevant code:

```rust
let mut js_objects = RootedVec::new();
 
for obj in self.iter() {
    let mut heap = Heap::default();
    let mut val = RootedValue::new(cx, UndefinedValue());
    obj.to_jsval(cx, val.handle_mut());
    heap.set(val.handle().get());
    js_objects.push(heap);
}
 
let values: Vec<JSVal> = js_objects.iter()
                                   .map(|heap| heap.handle().get())
                                   .collect();
<!-- Reviewable comment -K2c6vkzMM7WJETK27DJ:-844316405 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

**Comment by Ms2ger** [comment]
I wonder if we should have a `fn push(&mut self, Handle<T>)` for `RootedVec<Heap<T>>`. @nox?

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes +S-awaiting-answer 


---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oQrAIxHapd9WYkl81)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L820)):</sup>
I think the order here should be:
```rust
let mut val = RootedValue::new(cx, UndefinedValue());
obj.to_jsval(cx, val.handle_mut());
js_objects.push(Heap::default());
js_objects.iter_mut().last().unwrap().set(val.handle().get());
```


---

<sup>**[components/script/dom/bindings/conversions.rs, line 825 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oSkdTcZB3xOTp9amv)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L825)):</sup>
The fact that we need to create a new vector with bare Value types gives me pause. I think we need to sort out whether we should be using `RootedVec<Value>` or `RootedVec<Heap<Value>>` before I can finish this review.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 836 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oRmECIvJw7RVczbHO)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L836)):</sup>
`&*js_array` instead of transmuting, please.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by ecoal95** [comment]
Review status: 4 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oQrAIxHapd9WYkl81)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L820)):</sup>
I changed the implementation to take rid of this, using `JS_NewArrayObject1` (the overload with predefined length instead of values) and `JS_DefineElement` (to set numeric properties).

This also solves the double vector issue, and the `RootedVec`/`Heap` stuff.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 825 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oSkdTcZB3xOTp9amv)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L825)):</sup>
See above.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 836 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oRmECIvJw7RVczbHO)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L836)):</sup>
Done.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by jdm** [comment]
Great; only nits remaining!
-S-needs-rebase -S-awaiting-review -S-awaiting-answer +S-needs-code-changes 


---

Reviewed 3 of 3 files at r5.
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oQrAIxHapd9WYkl81)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L820)):</sup>
Great solution :)


---

<sup>**[components/script/dom/bindings/conversions.rs, line 43 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rYxOh_Tf-JPCeO3u8)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L43)):</sup>
js::JSPROP_ENUMERATE


---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZQzumQpK4c7hS4km)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L820)):</sup>
No need for `get()` here; `Handle<T>` dereferences to `&T`.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 822 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZ3iZdFu5LAggIq05)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L822)):</sup>
for (index, obj) in self.iter().enumerate() {


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by jdm** [comment]
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by ecoal95** [comment]
Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZQzumQpK4c7hS4km)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L820)):</sup>
Done.


---

<sup>**[components/script/dom/bindings/conversions.rs, line 822 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZ3iZdFu5LAggIq05)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L822)):</sup>
Good catch (sorry, I have little time and I tried to do it as quickly as possible).


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by jdm** [comment]
@bors-servo: r+


---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)

**Comment by jdm** [comment]
@bors-servo: retry
* #8468
